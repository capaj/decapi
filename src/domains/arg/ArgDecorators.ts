import { argRegistry } from './registry'
export { compileFieldArgs } from './compiler'
import { IArgOptions, defaultArgOptions } from './options'

export function Arg(options: IArgOptions = {}): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    argRegistry.set(target.constructor, [fieldName, argIndex], {
      ...defaultArgOptions,
      ...options,
      argIndex
    })
  }
}

/**
 * a shorthand for @Arg({isNullable: true})
 */
export function ArgNullable(options: IArgOptions = {}): ParameterDecorator {
  return Arg({ ...options, isNullable: true })
}
