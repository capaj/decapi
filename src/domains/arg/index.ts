import { argRegistry } from './registry'
export { compileFieldArgs } from './compiler'
import { ArgOptions, defaultArgOptions } from './options'

export function Arg(options: ArgOptions = {}): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    argRegistry.set(target.constructor, [fieldName, argIndex], {
      ...defaultArgOptions,
      ...options,
      argIndex,
    })
  }
}
