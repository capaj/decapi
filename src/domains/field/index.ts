import { fieldsRegistry, FieldInnerConfig } from './registry'

export {
  FieldInnerConfig,
  fieldsRegistry,
  queryFieldsRegistry,
} from './registry'
export { compileAllFields, compileFieldConfig } from './compiler/index'
export { FieldError } from './error'

export interface FieldOptions {
  description?: string
  type?: any
  name?: string
  isNullable?: boolean
}

export function Field(options?: FieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: true,
      ...options,
    }
    const existingField = fieldsRegistry.get(
      targetInstance.constructor,
      fieldName,
    )
    if (existingField) {
      throw new TypeError(
        `Field "${fieldName}" on class ${
          targetInstance.constructor.name
        } cannot be registered-it's already registered as type ${
          existingField.type.name
        }`,
      )
    }

    fieldsRegistry.set(targetInstance.constructor, fieldName, {
      ...finalConfig,
    })
  }
}
