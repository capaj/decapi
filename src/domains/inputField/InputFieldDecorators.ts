import { inputFieldsRegistry, IFieldInputInnerConfig } from './registry'

export {
  IFieldInputInnerConfig as FieldInputInnerConfig,
  inputFieldsRegistry
} from './registry'
export {
  compileAllInputFields,
  compileInputFieldConfig
} from './compiler/index'
export { InputFieldError } from './error'

export interface IInputFieldOptions {
  description?: string
  defaultValue?: any
  type?: any
  name?: string
  isNullable?: boolean
}

export function InputField(options?: IInputFieldOptions): PropertyDecorator {
  return (targetInstance: typeof Object, fieldName: string) => {
    if (
      options &&
      options.hasOwnProperty('type') &&
      options.type === undefined
    ) {
      console.log(
        'This can happens when a circular dependency is present. Wrap your explicit type in an arrow function to avoid this problem.'
      )
      throw new TypeError(
        `InputField "${fieldName}" explicit type is "undefined" on InputObjectType ${targetInstance.constructor}`
      )
    }

    const finalConfig: IFieldInputInnerConfig = {
      property: fieldName,
      name: fieldName,
      ...options
    }

    inputFieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig)
  }
}

/**
 * a shorthand for @InputField({isNullable: true})
 */
export function InputFieldNullable(
  options?: IInputFieldOptions
): PropertyDecorator {
  return InputField({ ...options, isNullable: true })
}
