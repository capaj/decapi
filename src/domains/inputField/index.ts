import { inputFieldsRegistry, FieldInputInnerConfig } from './registry'

export { FieldInputInnerConfig, inputFieldsRegistry } from './registry'
export {
  compileAllInputFields,
  compileInputFieldConfig
} from './compiler/index'
export { InputFieldError } from './error'

export interface InputFieldOptions {
  description?: string
  defaultValue?: any
  type?: any
  name?: string
  isNullable?: boolean
}

export function InputField(options?: InputFieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInputInnerConfig = {
      property: fieldName,
      name: fieldName,
      ...options
    }

    inputFieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig)
  }
}
