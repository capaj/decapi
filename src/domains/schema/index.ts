export {
  schemaRootsRegistry,
  mutationFieldsRegistry,
  queryFieldsRegistry,
} from './registry'
import { schemaRootsRegistry, SchemaRootConfig } from './registry'

// import { compileSchema } from './compiler';
export { compileSchema } from './compiler'
export { Query, Mutation } from './rootFields'
export { isSchemaRoot, getSchemaRootInstance } from './services'

export function SchemaRoot(config: SchemaRootConfig = {}): ClassDecorator {
  return (target) => {
    schemaRootsRegistry.set(target, config)
  }
}
