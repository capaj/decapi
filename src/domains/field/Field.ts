import { fieldsRegistry, IFieldInnerConfig } from './registry'
import { rootFieldTypes } from '../schema/rootFields'

export {
  IFieldInnerConfig as FieldInnerConfig,
  fieldsRegistry,
  queryFieldsRegistry
} from './registry'
export { compileAllFields, compileFieldConfig } from './compiler/fieldCompiler'
export { FieldError } from './error'

export interface IFieldOptions extends IFieldOptionsBase {
  isNullable?: boolean
}

export interface IFieldOptionsBase {
  description?: string
  rootFieldType?: rootFieldTypes
  type?: any
  castTo?: any
  onlyDecoratedArgs?: boolean
  name?: string
}

export function Field(options?: IFieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    if (
      options &&
      options.hasOwnProperty('type') &&
      options.type === undefined
    ) {
      console.log(
        'This usually happens when a circular dependency is present. Wrap your explicit type in an arrow function to avoid this problem.'
      )
      throw new TypeError(
        `Field "${fieldName}" on ${targetInstance.constructor} got an "undefined" as explicit type`
      )
    }
    const finalConfig: IFieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: true,
      ...options
    }
    const existingField = fieldsRegistry.get(
      targetInstance.constructor,
      fieldName
    )

    if (existingField) {
      if (!options.rootFieldType) {
        throw new TypeError(
          `Field "${fieldName}" on class ${targetInstance.constructor.name} cannot be registered-it's already registered as type ${existingField.type.name}`
        )
      }
      if (options.rootFieldType === existingField.rootFieldType) {
        throw new TypeError(
          `Root field "${fieldName}" on schema class ${targetInstance.constructor.name} cannot be registered as a ${existingField.rootFieldType}-it's already registered`
        )
      }
    }

    fieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig)
  }
}

/**
 * alias to help define array returning resolvers less verbosely
 */
export function ArrayField(options?: IFieldOptionsBase): PropertyDecorator {
  const typeOrCastTo = options ? options.type || options.castTo : null

  const optionsPassed = {
    ...options,
    isNullable: false
  }
  if (typeOrCastTo) {
    optionsPassed.type = Array.isArray(typeOrCastTo)
      ? typeOrCastTo
      : [typeOrCastTo]
  }
  return Field(optionsPassed)
}
