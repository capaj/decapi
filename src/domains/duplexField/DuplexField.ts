import { fieldsRegistry, IFieldInnerConfig } from '../field/registry'
import {
  inputFieldsRegistry,
  IFieldInputInnerConfig
} from '../inputField/registry'
import { IArrayFieldOptions } from '../field/Field'

export { FieldError } from './error'

interface IDuplexFieldOptions extends IArrayFieldOptions {
  inputNullable?: boolean
  itemCast?: any
}

export function DuplexField(options?: IDuplexFieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    let isNullable = true
    let inputNullable = true
    if (options) {
      if (options.isNullable !== undefined) {
        isNullable = options.isNullable
      }
      if (options.inputNullable !== undefined) {
        inputNullable = options.inputNullable
      }
      delete options.isNullable
      delete options.inputNullable
    }

    let type 
    if (options?.itemCast) {
      type = [options.itemCast]
    }

    const finalInputConfig: IFieldInputInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: inputNullable,
      type,
      ...options
    } as any

    const finalConfig: IFieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable,
      ...options
    } as any
    fieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig)

    inputFieldsRegistry.set(
      targetInstance.constructor,
      fieldName,
      finalInputConfig
    )
  }
}

export function DuplexArrayField(
  options?: IArrayFieldOptions
): PropertyDecorator {
  const typeOrCastTo = options?.itemCast ?? null

  if (!typeOrCastTo) {
    throw new TypeError(
      'ArrayField must have an explicit itemType or itemCast config'
    )
  }
  return DuplexField(options)
}
