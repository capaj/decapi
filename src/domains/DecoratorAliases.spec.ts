import { printSchema } from 'graphql'
import { DuplexObjectType, SchemaRoot, compileSchema } from '../index'

import { ArgNullable } from './arg/ArgDecorators'
import { QueryAndMutation } from './schema/rootFields'

import { InputFieldNullable } from './inputField/InputFieldDecorators'
import { DuplexField } from './duplexField/DuplexField'
import { ArrayField } from './field/Field'

describe.only('decorator aliases', () => {
  it('should compile', async () => {
    @DuplexObjectType()
    class Bar {
      @InputFieldNullable()
      foo2: string
      @DuplexField()
      foo(@ArgNullable() a: string): string {
        return a
      }
    }

    @SchemaRoot()
    class FooSchema {
      @QueryAndMutation()
      output(): Bar {
        return new Bar()
      }
      @QueryAndMutation()
      echo(input: Bar): Bar {
        return new Bar()
      }
    }
    const schema = compileSchema(FooSchema)
    expect(printSchema(schema)).toMatchSnapshot()
  })

  describe('ArrayField', () => {
    it.only('should define the field correctly', async () => {
      @DuplexObjectType()
      class Bar {
        @ArrayField()
        foo2: string
        @ArrayField({ type: () => String })
        foo(@ArgNullable() a: string): string[] {
          return []
        }
        @ArrayField({ type: Number })
        foo3() {
          return [1, 2, 3]
        }
      }

      @SchemaRoot()
      class FooSchema {
        @QueryAndMutation()
        output(): Bar {
          return new Bar()
        }
      }
      const schema = compileSchema(FooSchema)
      expect(printSchema(schema)).toMatchSnapshot()
    })
  })
})
