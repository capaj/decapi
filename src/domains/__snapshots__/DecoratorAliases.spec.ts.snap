// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorator aliases ArrayField should define the field correctly 1`] = `
"type Bar {
  foo2: String!
  foo(a: String): [String!]!
  foo3: [Float!]!
}

type Mutation {
  output: Bar
}

type Query {
  output: Bar
}
"
`;

exports[`decorator aliases should compile 1`] = `
"type Bar {
  foo(a: String): String
}

input BarInput {
  foo2: String
  foo: String
}

type Mutation {
  output: Bar
  echo(input: BarInput!): Bar
}

type Query {
  output: Bar
  echo(input: BarInput!): Bar
}
"
`;
